larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,892 @@
+
+bool TrackInEMShowerAlgorithm::IsInLineSegment(const CartesianVector &lowerBoundary, const CartesianVector &upperBoundary, const CartesianVector &point) const
+{
+    const float gradient = (-1.0)*(upperBoundary.GetX() - lowerBoundary.GetX()) / (upperBoundary.GetZ() - lowerBoundary.GetZ());
+    const float xPointOnUpperLine((point.GetZ() - upperBoundary.GetZ() + gradient*upperBoundary.GetX())/gradient);
+    const float xPointOnLowerLine((point.GetZ() - lowerBoundary.GetZ() + gradient*lowerBoundary.GetX())/gradient);
+
+    const CartesianVector upper(xPointOnUpperLine, 0.f, point.GetZ());
+    const CartesianVector lower(xPointOnLowerLine, 0.f, point.GetZ());
+
+    if ((point.GetX() > xPointOnUpperLine) && (point.GetX() > xPointOnLowerLine))
+        return false;
+
+    if ((point.GetX() < xPointOnUpperLine) && (point.GetX() < xPointOnLowerLine))
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::RefineTracks(ClusterAssociation &clusterAssociation, TwoDSlidingFitResultMap &microSlidingFitResultMap,
+    TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector, CaloHitVector &extrapolatedCaloHitVector) const
+{
+    clusterAssociation.SetUpstreamCluster(this->RefineTrack(clusterAssociation.GetUpstreamCluster(), clusterAssociation.GetUpstreamMergePoint(), extrapolatedCaloHitVector, true,
+        microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+    clusterAssociation.SetDownstreamCluster(this->RefineTrack(clusterAssociation.GetDownstreamCluster(), clusterAssociation.GetDownstreamMergePoint(), extrapolatedCaloHitVector, false,
+        microSlidingFitResultMap, macroSlidingFitResultMap, clusterVector));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const Cluster* TrackInEMShowerAlgorithm::RefineTrack(const Cluster *const pCluster, const CartesianVector &splitPosition, const CaloHitVector &extrapolatedCaloHitVector,
+    const bool isUpstream, TwoDSlidingFitResultMap &microSlidingFitResultMap, TwoDSlidingFitResultMap &macroSlidingFitResultMap, ClusterVector &clusterVector) const
+{
+    float rL(0.f), rT(0.f);
+    const TwoDSlidingFitResult &microFitResult(microSlidingFitResultMap.at(pCluster));
+    microFitResult.GetLocalPosition(splitPosition, rL, rT);
+
+    const TwoDSlidingFitResult macroFitResult(macroSlidingFitResultMap.at(pCluster));
+    CartesianVector averageDirection(0.f, 0.f, 0.f);
+    macroFitResult.GetGlobalDirection(macroFitResult.GetLayerFitResultMap().begin()->second.GetGradient(), averageDirection);
+
+    const float clusterGradient(averageDirection.GetZ()/averageDirection.GetX());           
```

Comment:
I've changed the code so that if the cluster direction has a zero x component the cluster is marked as vertical. Then in the following the gradient and intercept correspond to a straight line fit in which the gradient corresponds to dx/dz (which is 0 for this case) so instead the hits are split into a left and right cluster